### PR TITLE
Add a bunch of guides for testing

### PR DESCRIPTION
**Pull request by gaganawhad**

### CODE REVIEW

File: best-practices/README.md [hunk 1]
```diff
@@ -82,12 +82,21 @@ Rails
 Testing
 -------
 
-* Avoid using instance variables in tests.
-* Disable real HTTP requests to external services with
-  `WebMock.disable_net_connect!`.
+* Follow guidelines in [Better Specs][better-specs]
+* Look to your team to make judgment calls. Your team is your best friend in helping you make them.
+* Break the rules when it saves money!
```
Comment (joeosburn-desiringgod) [inline best-practices/README.md]: @gaganawhad This is not my favorite way to make this point.

File: best-practices/README.md [hunk 2]
```diff
@@ -82,12 +82,21 @@ Rails
...
+* Write tests for non-trivial code. This saves us money. 
+* Don't write tests for trivial code. This also saves us money.
+* Make good judgements on what is trivial and what is not. Not everything that looks trivial, is so. Simple is not trivial.
+* Test only happy paths in acceptance tests. Acceptances tests are end-to-end tests. 
+  They are very expensive! Testing happy paths gives you enough assurance that all layers of code are working!
+* Move exceptions, non-happy paths, complicated logic to classes and unit test them.
+* Avoid using instance variables in ruby unit tests.
 * Don't test private methods.
-* Use acceptance tests to execute the entire app.
+* Avoid writing tests purely for the reason of specification / documentation on libraires that are familiar and reliable.
```
Comment (joeosburn-desiringgod) [inline best-practices/README.md]: @gaganawhad Small typo here.

File: best-practices/README.md [hunk 3]
```diff
@@ -82,12 +82,21 @@ Rails
...
-* Disable real HTTP requests to external services with
-  `WebMock.disable_net_connect!`.
+* Follow guidelines in [Better Specs][better-specs]
+* Look to your team to make judgment calls. Your team is your best friend in helping you make them.
+* Break the rules when it saves money! 
+* Write tests for non-trivial code. This saves us money. 
+* Don't write tests for trivial code. This also saves us money.
+* Make good judgements on what is trivial and what is not. Not everything that looks trivial, is so. Simple is not trivial.
+* Test only happy paths in acceptance tests. Acceptances tests are end-to-end tests. 
+  They are very expensive! Testing happy paths gives you enough assurance that all layers of code are working!
```
Comment (joeosburn-desiringgod) [inline best-practices/README.md]: @gaganawhad The exclamation mark here is not necessary.

File: best-practices/README.md [hunk 4]
```diff
@@ -82,12 +82,21 @@ Rails
...
+* Move exceptions, non-happy paths, complicated logic to classes and unit test them.
```
Comment (joeosburn-desiringgod) [inline best-practices/README.md]: @gaganawhad I'm not sure if I follow this sentence 100%. Is this what you meant to say?

File: best-practices/README.md (hunk 3 above)
Comment (benhutton) [inline best-practices/README.md]: should there be a `*` at the beginning of this line?

File: best-practices/README.md (hunk 1 above)
Comment (gaganawhad) [inline best-practices/README.md]: @joeosburn-desiringgod any recommendations?

File: best-practices/README.md (hunk 3 above)
Comment (gaganawhad) [inline best-practices/README.md]: @benhutton @joeosburn-desiringgod That's not a separate point hence I didn't add in a bullet. It's just a way I am maintaining some limit on the line length. See https://github.com/desiringgod/guides/blob/eb9a68d87ac14d93544657e5fe4414f3472fef9a/best-practices/README.md

File: best-practices/README.md (hunk 4 above)
Comment (gaganawhad) [inline best-practices/README.md]: @joeosburn-desiringgod Makes sense! I'll improve it.

**Comment by gaganawhad** [comment]
@joeosburn-desiringgod Take another look!

File: best-practices/README.md (hunk 1 above)
Comment (joeosburn-desiringgod) [inline best-practices/README.md]: How about:

`The rules are intended to serve us, not rule us. If a rule is not benefiting us, it can be circumvented. Please consider submitting a PR to update the guides repository if you find yourself in this situation.`

File: best-practices/README.md (hunk 3 above)
Comment (joeosburn-desiringgod) [inline best-practices/README.md]: I'm fine without them being a separate bullet point but I'd still say we should kill the extra exclamation mark.

File: best-practices/README.md [hunk 5]
```diff
@@ -82,12 +82,21 @@ Rails
...
-  `WebMock.disable_net_connect!`.
+* Follow guidelines in [Better Specs][better-specs]
+* Look to your team to make judgment calls. Your team is your best friend in helping you make them.
+* Break the rules when it saves money! 
+* Write tests for non-trivial code. This saves us money. 
+* Don't write tests for trivial code. This also saves us money.
+* Make good judgements on what is trivial and what is not. Not everything that looks trivial, is so. Simple is not trivial.
+* Test only happy paths in acceptance tests. Acceptances tests are end-to-end tests. 
+  They are very expensive! Testing happy paths gives you enough assurance that all layers of code are working!
+* Avoid testing exceptions, edge cases, complicated logic in acceptances tests. Prefer unit tests to test them.
```
Comment (joeosburn-desiringgod) [inline best-practices/README.md]: @gaganawhad This is good, the only potential difficulty here is "complicated" logic. I think I get what you mean. "complicated" might be the best word, but I'm not sure. What about:

`Avoid testing exceptions, edge cases, and convoluted logic in acceptances tests. Prefer unit tests to test them.`

File: best-practices/README.md (hunk 1 above)
Comment (gaganawhad) [inline best-practices/README.md]: @joeosburn-desiringgod That's definitely the spirit of that statement.

Regarding submitting the PR we have a general guide for all guides here: https://github.com/desiringgod/guides/blob/master/README.md#contributing. Does that help? 

I wonder if there is a problem with the use of the word 'rule'. How about I make it... 
> Break the guides when it saves money!
will that help? 

Any thoughts here @benhutton ?

File: best-practices/README.md (hunk 5 above)
Comment (gaganawhad) [inline best-practices/README.md]: @joeosburn-desiringgod I am happy to make that word change.

**Comment by gaganawhad** [comment]
Allright @joeosburn-desiringgod ... I have either made updates or written comments on the PR here, so it is ready for you.

File: best-practices/README.md (hunk 1 above)
Comment (benhutton) [inline best-practices/README.md]: I would just include the first half of Joe's suggestion, without the PR sentence.

**Comment by gaganawhad** [comment]
@benhutton @joeosburn-desiringgod All the changes are done. I can merge once you approve.
